### ISSUE TITLE
IRC over QUIC as a standard

### ISSUE THREAD

**Issue by TaaviE** [ircv3/ircv3-ideas#28]
I think it's worth starting to think about providing IRC over QUIC for these reasons:
   - Mandatory strong encryption - This basically provides HSTS but for IRC, helping to somewhat solve this: https://github.com/ircv3/ircv3-specifications/issues/300
   - In theory provides lower overhead, latency and reduces transfer sizes - 0RTT is also really nice
   - Could in theory provide smooth connection migration when an IRC server goes down
   - Has a lot of development power behind it from big corporations
   - We'll most certainly see Websockets over QUIC(or HTTP/3) so that would be easier to provide when there's an existing QUIC capability in the server

What do you think? Should it maybe even be called IRCv4 :P?

**Comment by grawity** [comment]
> Maybe with just TCP but with TCP+TLS it's way more complex than what QUIC offers though?

As far as I know, the transport is nearly irrelevant to the problem – most of the complexity is on the application layer anyway (getting the new server to recognize you, rebuild state, update routing for the nickname across the network). But do you have a specific mechanism in mind already? That is, what precisely does QUIC offer for this feature?

For everything else, QUIC would do the job, but it doesn't sound like it needs a specific "IRC over QUIC" profile, just a generic "QUIC as TLS substitute", same as SMTP or IMAP...

> Has a lot of development power behind it from big corporations

Do any of them care about non-HTTP applications?

**Comment by jwheare** [comment]
Moved to the ideas repo.

**Comment by ghost** [comment]
Id rather go for SCTP instead of QUIC.

**Comment by TaaviE** [comment]
SCTP gets mangled by middleboxes and will not have such a good support as QUIC/HTTP3 does in various (standard)libraries.

**Comment by NetworkJack2** [comment]
I don't think it should be mandatory. Perhaps a feature where is mandatory for being labeled IRCv3, but also consider that unencrypted IRC may be used for networks that are already private/encrypted and is both redundant and hard to implement a Key Infrastructure.

VPNs, and darknets like TOR and I2P come to mind.

**Comment by bortzmeyer** [comment]
> SCTP gets mangled by middleboxes 

Then, SCTP-over-UDP (RFC 6951), for the same reason QUIC uses UDP.

**Comment by sadiepowell** [comment]
The main reason i'm personally uninterested in SCTP is that there's no support in either Windows or macOS (which has it IPPROTO_SCTP in the headers but creating a socket using it results in an error) which makes it not very useful for the average user.

**Comment by ghost** [comment]
We should then not blame SCTP but Microsoft and Apple for that.

**Comment by sadiepowell** [comment]
You can blame whoever you want but that doesn't change the reality that features which 90% of potential users can't use are useless.

If Google couldn't get OS vendors to care about SCTP then we won't be able to.

**Comment by ghost** [comment]
There is a proper driver which can be installed for OSX and one for windows, but I heard its not in a good state.
Maybe we should get more SCTP applications first and then demand OS vendors to add SCTP drivers.

**Comment by TaaviE** [comment]
IMO SCTP talk should be left to some other issue.

**Comment by justjanne** [comment]
Please move SCTP discussion over here: https://github.com/ircv3/ircv3-ideas/issues/62

**Comment by justjanne** [comment]
The current QUIC interop situation is this: https://docs.google.com/spreadsheets/d/1D0tW89vOoaScs3IY9RGC0UesWGAwE6xyLk0l4JtvTVg/edit#gid=117825384

The current draft for IETF QUIC is still moving relatively **quic**kly, so it makes sense to be cautious and wait for consensus before standardising this, but IMO it'd be interesting to experiment with the current [14th draft](https://github.com/quicwg/base-drafts/wiki/14th-Implementation-Draft) against which implementations are currently slowly converging.

**Comment by TheDecryptor** [comment]
QUICv1 has been finished for a while at this point (A small v2 is currently being worked on), support should be good enough that devs shouldn't expect to run into issues with functionality or compatibility
